### PR TITLE
mc_att_control: move rate control to RateControl class

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by the proposed pull request**
This makes the controller more modular, more readable, unit-testable and hence better maintainable. It heavily simplifies any architecture changes like the separate rate control module proposed in #12225.

**Describe your preferred solution**
Like with the attitude control part of this module (#11308) I pulled out all the angular rate control related calculations into a class named `RateControl` with a clear interface.

Additional improvments:
- Added d term calculation guard if dt is zero to make the basic all zero test pass.
https://github.com/PX4/Firmware/compare/rate-control-class?expand=1#diff-cebc2f2b5e55acab8a0b02a1b8b3e6bfR89-R91
- Improved mixer saturation interface to save these rolled out checks for each axis
https://github.com/PX4/Firmware/compare/rate-control-class?expand=1#diff-81c4f50bdf837c4183b5d3cb81eb40eaL432-L441
- split out integral update into a separate function to increase readability
https://github.com/PX4/Firmware/compare/rate-control-class?expand=1#diff-cebc2f2b5e55acab8a0b02a1b8b3e6bfR99-R102

**Test data / coverage**
SITL jMAVsim and gazebo normal test flights. No case/configuration specific tests yet.

### CODE REVIEW

File: src/modules/mc_att_control/RateControl/RateControl.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	_mixer_saturation_positive[2] = status.flags.yaw_pos;
+	_mixer_saturation_negative[0] = status.flags.roll_neg;
+	_mixer_saturation_negative[1] = status.flags.pitch_neg;
+	_mixer_saturation_negative[2] = status.flags.yaw_neg;
+}
+
+Vector3f RateControl::update(const Vector3f rate, const Vector3f rate_sp, const float dt, const bool landed,
+			     const float thrust_sp)
+{
+	Vector3f gain_p_tpa = _gain_p.emult(tpa_attenuations(_tpa_breakpoint(0), _tpa_rate(0), thrust_sp));
```
Comment (dagar) [inline src/modules/mc_att_control/RateControl/RateControl.cpp]: One little optimization we should add is only updating TPA on thrust setpoint changes. At 1 kHz that alone saves more than half a percent of total cpu usage.

Comment (MaEtUgR) [reply src/modules/mc_att_control/RateControl/RateControl.cpp]: Actually I plan to get rid of TPA all together. I just didn't want to pack that into this refactoring pr.

TPA is the dirty solution to the more quadratic than linear thrust output function of your ESC, motor, propeller. But it's far less acurate and up to tuning of each axis which basically no one does.

I think https://docs.px4.io/master/en/advanced_config/parameter_reference.html#THR_MDL_FAC the mixer/output level compensation with lnear combination of linear and quadratic function is the far better approach and I'm only using that one. Here's also a good example: https://discuss.px4.io/t/thrust-curve-compensation-through-thr-mdl-fac/13207

What we need to make sure is that this way of compensating is well documented and then I don't see a reason to keep TPA.

Comment (dagar) [reply src/modules/mc_att_control/RateControl/RateControl.cpp]: Nice, I was wondering why we had both.

Comment (MaEtUgR) [reply src/modules/mc_att_control/RateControl/RateControl.cpp]: >Nice, I was wondering why we had both.

Someone added it, someone used it, the other method was not documented for a long time. Noone removed it. I'll do that.

Now it's actually nicely documented 🎉  https://docs.px4.io/v1.9.0/en/config_mc/pid_tuning_guide_multicopter.html#thrust_curve
We could add a complete example for an even lower barrier.

**Comment by dannyfpv** [comment]
tested on pixhawk4 v5 f-450 
Modes Tested
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.
- Notes
No issues noted, good flight in general.

logs:
fc1:       https://review.px4.io/plot_app?log=3cf50ba0-e82f-4f51-95c1-0ab90626145f  
fc2:       https://review.px4.io/plot_app?log=505e5226-365e-4200-a6b8-4dc2b59ffa5d
fc3:       https://review.px4.io/plot_app?log=e85bd82b-0da8-4684-ae97-25792482d1c8
fc4        https://review.px4.io/plot_app?log=ffe9182c-49b5-4fa2-855c-df04ccd814d8

**Comment by jorge789** [comment]
### Tested on PixRacer V4:

**Flight Card 1**
Modes Tested:
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

**Notes:**
No issues were noted, good flight in general.

**Logs:**
https://review.px4.io/plot_app?log=f74c845e-d992-4246-b827-cd4ee7c36fac

https://review.px4.io/plot_app?log=cdcfdb80-e5bf-4874-90b9-2c01974fd208

**Flight Card 2**

**Modes Tested**
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
- Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed as expected.
Good flight in general.

**Note:**
No issues were noted, good flight in general.

**Logs:**
https://review.px4.io/plot_app?log=d482d9d8-7027-469a-b582-eb1d5f652686

https://review.px4.io/plot_app?log=0c511d73-bf96-49e3-9d41-c7f216a1d8a8

**Flight Card 3**
**Modes Tested**
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.
Good flight in general.

**Note:**
No issues were noted, good flight in general.

Logs:
https://review.px4.io/plot_app?log=1abb3f63-feb4-4d69-af11-11f9b423a13c

https://review.px4.io/plot_app?log=8a5f4122-4fd9-4a6f-9ef3-3d16c6e5c3ed

**Flight Card 4**
**Modes Tested**
Fail safe:The radio control turned off and the vehicle returned and landed correctly.
Tested good.

Logs:
https://review.px4.io/plot_app?log=651a80b2-2704-4f7f-825f-42b050bd3526

https://review.px4.io/plot_app?log=3fb9e420-64e1-422f-bf70-74aa2a7220e1

**Comment by dannyfpv** [comment]
tested on pixhawkprov4  f-450
Modes Tested
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.

Notes
No issues noted, good flight in general.
logs:
fc1: https://review.px4.io/plot_app?log=383f7723-c369-4cd8-8476-0edb1537c036
fc2: https://review.px4.io/plot_app?log=b47bb278-327d-4d07-8f15-5a45b12bfc1d
fc3: https://review.px4.io/plot_app?log=1d9981a4-ced2-40a1-93a4-25c80fb3f5b7
fc4 https://review.px4.io/plot_app?log=c561abe0-31e0-4833-9128-678d917b7677

**Comment by MaEtUgR** [comment]
@bresch About the discussion we had yesterday: I double-checked and the gains stay the same so the fix in https://github.com/PX4/Firmware/commit/e9ab6a75bada21ff907ce7b76c9fe42a59b87585#diff-81c4f50bdf837c4183b5d3cb81eb40ea is preserved by scaling all the gains including the integrator when passing the parameters to the setter here: https://github.com/PX4/Firmware/pull/13173/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR102-R104

**Pull request merged**